src/ucai/core/databricks.py

```diff
@@ -111,6 +125,46 @@ def extract_function_name(sql_body: str) -> str:
     )
 
 
+def retry_on_session_expiration(func):
+    """
+    Decorator to retry a method upon session expiration errors with exponential backoff.
+    """
+    max_attempts = int(UCAI_DATABRICKS_SESSION_RETRY_MAX_ATTEMPTS.get())
+
+    @functools.wraps(func)
+    def wrapper(self, *args, **kwargs):
+        for attempt in range(1, max_attempts + 1):
+            try:
+                return func(self, *args, **kwargs)
+            except Exception as e:
+                error_message = str(e)
+                if SESSION_EXCEPTION_MESSAGE in error_message:
+                    if not self._is_default_client:
+                        refresh_message = f"Failed to execute {func.__name__} due to session expiration. Unable to automatically refresh session when using a custom client."
+                        _logger.error(refresh_message)
+                        raise RuntimeError(refresh_message) from e
+
+                    if attempt < max_attempts:
+                        delay = min(
+                            SESSION_RETRY_BASE_DELAY * (2 ** (attempt - 1)),
+                            SESSION_RETRY_MAX_DELAY,
+                        )
+                        _logger.warning(
+                            f"Session expired. Attempt {attempt} of {max_attempts}. Refreshing session and retrying after {delay} seconds..."
+                        )
+                        self.refresh_client_and_session()
+                        time.sleep(delay)
+                        continue
+                    else:
+                        refresh_failure_message = f"Failed to execute {func.__name__} after {max_attempts} attempts due to session expiration."
+                        _logger.error(refresh_failure_message)
```

Comment:
Same here, seems no need

Reply:
As a final statement at the end of an unrecoverable error, logging makes zero sense :) removed!